File: library/src/main/java/com/applandeo/materialcalendarview/utils/SelectedDay.java
package com.applandeo.materialcalendarview.utils;

import android.view.View;

import java.util.Calendar;

/**
 * This helper class represent a selected day when calendar is in a picker date mode.
 * It is used to remember a selected calendar cell.
 * <p>
 * Created by Mateusz Kornakiewicz on 23.05.2017.
 */

public class SelectedDay {
    private View mView;
    private Calendar mCalendar;

    public SelectedDay(Calendar calendar) {
        mCalendar = calendar;
    }

    /**
     * @param view     View representing selected calendar cell
     * @param calendar Calendar instance representing selected cell date
     */
    public SelectedDay(View view, Calendar calendar) {
        mView = view;
        mCalendar = calendar;
    }

    /**
     * @return View representing selected calendar cell
     */
    public View getView() {
        return mView;
    }

    public void setView(View view) {
        mView = view;
    }

    /**
     * @return Calendar instance representing selected cell date
     */
    public Calendar getCalendar() {
        return mCalendar;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof SelectedDay) {
            return getCalendar().equals(((SelectedDay) obj).getCalendar());
        }

        if(obj instanceof Calendar){
            return getCalendar().equals(obj);
        }

        return super.equals(obj);
    }
}

File: library/src/main/java/com/applandeo/materialcalendarview/utils/SelectedDay.kt
package com.applandeo.materialcalendarview.utils

import android.view.View
import java.util.*

/**
 * This helper class represent a selected day when calendar is in a picker date mode.
 * It is used to remember a selected calendar cell.
 *
 * @param calendar Calendar instance representing selected cell date
 * @param view     View representing selected calendar cell
 *
 * Created by Applandeo Team.
 */

data class SelectedDay @JvmOverloads constructor(
        val calendar: Calendar,
        var view: View? = null
) {

    override fun equals(other: Any?): Boolean {
        return when (other) {
            is SelectedDay -> calendar.isTheSameDay(other.calendar)
            is Calendar -> calendar.isTheSameDay(other)
            else -> super.equals(other)
        }
    }

    private fun Calendar.isTheSameDay(calendar: Calendar): Boolean =
            this.get(Calendar.DAY_OF_MONTH) == calendar.get(Calendar.DAY_OF_MONTH) &&
                    this.get(Calendar.MONTH) == calendar.get(Calendar.MONTH) &&
                    this.get(Calendar.YEAR) == calendar.get(Calendar.YEAR)
}